Step4

mz-bskoo opened the pull request:
리뷰 부탁드립니다.

mz-bskoo committed:
docs: 기능목록 정의

mz-bskoo committed:
feat: 자동차이름 입력기능 추가

mz-bskoo committed:
feat: 전진하는 자동차 출력할때 자동차 이름을 같이 출력

mz-bskoo committed:
refactor: 상수값 변수 사용

mz-bskoo committed:
feat: 우승자선정 및 화면출력

mz-bskoo committed:
test: 자동차이름 길이 테스트

- 자동차이름 최대 길이를 Car에서 처리

mz-bskoo committed:
refactor: 자동차 최대 이동거리 메서드 분리

yamsroun commented on src/main/java/com/megazone/goop/racingcar/logic/Car.java:
```diff
@@ -4,8 +4,19 @@ public class Car {
 
+    public Car(String name) {
+        this.name = name.substring(0, Math.min(name.length(), CAR_NAME_MAX_LENGTH));
+    }
```
요구사항 만족을 위해 최대 자릿수로 잘라내는거 좋네요~ 👍🏾👍🏾

현재 미션은 OOP 연습이므로 이렇게 한번 제안 드려봅니다.
자동차 이름 VO(예: `CarName`)를 만들어 보세요.
`Car` 객체는 `String name` 대신 `CarName name`를 갖게 되는 구조입니다.
그리고 자동차 이름과 관련된 규칙(로직)은 VO에서 처리하도록 수정해 주세요.

mz-bskoo replied:
CarName Class 를 생성해서 자동차 이름에 대한 유효성 체크를 하도록 변경하였습니다.

yamsroun commented on src/main/java/com/megazone/goop/racingcar/logic/Race.java:
```diff
@@ -11,23 +10,14 @@ public class Race {
 
-    public Race(ResultView resultView, RandomGenerator randomGenerator, int carCount) {
+    public Race(List<Car> cars, ResultView resultView, RandomGenerator randomGenerator) {
+        this.cars = cars;
```
`cars`를 외부에서 생성해서 주입받을 수도 있겠지만,
그런 경우 사이드-이펙트가 발생할 소지도 있습니다.

예를 들어, (외부에서 만들었으므로) `cars` 객체가 외부에 그대로 노출되기 때문에
외부에서 cars 객체를 변경할 수 있습니다. (`List` 자체 또는 `List` 요소인 `Car` 객체를 변경 가능)
그렇게 되면, Race 객체 내에서 진행하는 경주와 무관하게 누군가가(?) 결과를 조작할 수 있습니다.

1) 외부에서 cars를 입력받되 내부에서 복제하는 형태,
2) 또는, 기존처럼 내부에서 cars를 생성하는 형태로
수정이 필요해 보입니다.

mz-bskoo replied:
사용자 입력을 전달해서 Race 내부에서 cars 를 생성하도록 변경하였습니다.

yamsroun commented on src/main/java/com/megazone/goop/racingcar/logic/Race.java:
```diff
@@ -38,7 +28,7 @@ public void startRace(int roundCount) {
             cars.forEach(c -> c.move(randomGenerator.getRandomValue(MAX_RANDOM_VALUE)));
 
-            resultView.displayPositions(cars.stream().map(Car::getTotalDistance).toList());
+            resultView.displayPositions(cars);
```
결과를 표시하는 코드가 간결해졌네요.

이 부분도 위 코멘트와 비슷한 문제가 있습니다.
`cars` 객체를 다른 객체로 그대로 전달하는데, 전달 받은 객체(`resultView`)에서 이 `cars` 객체를 변경한다면,
`Race` 객체의 경주 결과가 조작될 수 있습니다.

mz-bskoo replied:
race에서 resultView의 메서드를 호출하지 않고, race 외부에서 resultView로 race 결과를 전달하는 형태로 수정하였습니다.

yamsroun commented on src/main/java/com/megazone/goop/racingcar/logic/Race.java:
```diff
@@ -48,9 +38,24 @@ public int getCurrentRound() {
 
+    public void endRace() {
+        List<Car> winners = cars.stream()
+            .filter(c -> c.getTotalDistance() >= getTopRecord())
```
`getTopRecord()`가 반복 호출되므로 스트림 처리도 반복됩니다.
최적화가 필요합니다.

yamsroun commented on src/main/java/com/megazone/goop/racingcar/logic/Car.java:
```diff
@@ -4,8 +4,19 @@ public class Car {
+    private static final int CAR_NAME_MAX_LENGTH = 5;
+
+    private final String name;
     private int totalDistance = 0;
```
`CarName` VO와 마찬가지로, 자동차 위치도 VO로 표현 가능합니다.
현재는 딱히 필요성이 보이지 않아 하지 않으셔도 됩니다.

mz-bskoo replied:
넵.

yamsroun commented on src/main/java/com/megazone/goop/racingcar/ui/InputView.java:
```diff
@@ -1,24 +1,30 @@
 
-        System.out.println("자동차 대수는 몇 대 인가요?");
-        carCount = scanner.nextInt();
+        System.out.println("경주할 자동차 이름을 입력하세요. (이름은 쉼표(,)를 기준으로 구분)");
+        String[] inNames = scanner.nextLine().split(CAR_NAME_DELIMITER);
```
이런 케이스도 문제가 없도록(?) 처리해 주세요.
<img width="514" alt="image" src="https://github.com/g-oop/goop-java-racingcar/assets/20655591/a551c720-eb59-481a-9f8f-8d7186368089">

yamsroun reviewed (changes requested):
몇 가지 코멘트 남겼습니다~ 😁😁😁
확인 후 처리 부탁 드립니다~ 🙏🏾🙏🏾🙏🏾

mz-bskoo committed:
refactor: 리뷰1차 반영

- CarName 클래스 생성해서 name 유효성 체크
- CarNameTest 코드 추가
- 자동차 이름 없는 경우 예외처리
- 사용자 입력을 RacePlan 으로 Race로 전달
- Race 클래스에서 자동차 생성하도록 수정
- 모든 라운드가 완료된 후 결과를 출력하도록 수정

mz-bskoo committed:
rename: 패키지 변경

mz-bskoo committed:
test: 자동차 이름없음 테스트 추가

yamsroun commented on src/main/java/com/megazone/goop/racingcar/car/CarName.java:
```diff
@@ -0,0 +1,17 @@
+public record CarName(String name) {
+    private static final int CAR_NAME_MAX_LENGTH = 5;
+
+    public CarName {
+        if (name == null || name.isBlank()) {
+            throw new IllegalArgumentException("자동차 이름이 설정되지 않았습니다.");
+        }
+        if (name.length() > CAR_NAME_MAX_LENGTH) {
+            throw new IllegalArgumentException(
+                MessageFormat.format("자동차 이름이 최대길이({})를 초과하였습니다.", CAR_NAME_MAX_LENGTH));
+        }
+    }
```
👍🏾

yamsroun commented on src/main/java/com/megazone/goop/racingcar/car/CarDistance.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.text.MessageFormat;
+
+public record CarDistance(String name, int position) {
```
`CarDistance` 객체에서 `String name` 필드는 적절하지 않아 보입니다.
객체가 표현하고자 하는 데이터와 무관한 필드로 보입니다.

그리고 Distance와 Position이라는 2가지 용어를 사용하셨는데 구분되는 이유가 있을까요?

yamsroun replied:
다시 보니, `CarDistance` 객체는 `Car` 객체의 VO가 아니군요.
`CarDistance` 객체의 정체를 명확히 정의할 필요가 있겠습니다.
현재는 DTO 목적과 출력 목적이 섞여있는 것으로 보입니다.
또한, 목적에 맞게 패키지 위치 조정도 필요합니다.

yamsroun commented on src/main/java/com/megazone/goop/racingcar/car/CarDistance.java:
```diff
@@ -0,0 +1,11 @@
+
+public record CarDistance(String name, int position) {
+
+    @Override
+    public String toString() {
+        return MessageFormat.format("{0} : {1}", name, "-".repeat(position));
+    }
```
`toString()`을 로그 용도가 아닌 목적으로 오버라이딩 하는 것은 주의가 필요합니다.
의견이 좀 나뉘지만, 저는 디버깅 목적으로만 사용하는 것이 속 편하고 사이드-이펙트가 없다고 생각합니다.
그리고 작성하신 출력 문자열은 VO의 역할이 아니라 ResultView의 역할에 가깝습니다.

<img width="738" alt="image" src="https://github.com/g-oop/goop-java-racingcar/assets/20655591/ec15d607-fa86-4681-9017-cfb2a96e27eb">

yamsroun commented on src/main/java/com/megazone/goop/racingcar/race/RacePlan.java:
```diff
@@ -0,0 +1,11 @@
+import com.megazone.goop.racingcar.car.CarName;
+
+public record RacePlan(
+    List<CarName> CarNames,
```
파라미터 변수명이 대문자로 시작하네요.😅

yamsroun commented on src/main/java/com/megazone/goop/racingcar/race/Race.java:
```diff
@@ -0,0 +1,73 @@
+            .toList();
+    }
+
+    private void readyRace() {
+        this.currentRound = 0;
+
+        cars.forEach(Car::resetPosition);
+    }
```
`private` 메서드는 호출하는 메서드 직후에 배치하는 것이 좋습니다.

yamsroun replied:
`roundRecords`도 초기화 해야하지 않나요?

yamsroun commented on src/main/java/com/megazone/goop/racingcar/ui/ResultView.java:
```diff
@@ -2,14 +2,28 @@
-    public void displayPositions(List<Integer> positions) {
-        positions.forEach(n -> System.out.println("-".repeat(n)));
-        System.out.println();
+    private void displayCarPositions(List<CarDistance> carDistances) {
+        carDistances.forEach(d ->
+            System.out.println(d.toString())
+        );
```
여기도 메서드 배치 위치가 호출하는 측에 가까우면 좋겠습니다.

yamsroun commented on src/test/java/com/megazone/goop/racingcar/util/RandomGeneratorTest.java:
```diff
@@ -1,7 +1,9 @@
+
 import static org.assertj.core.api.Assertions.assertThat;
 
 class RandomGeneratorTest {
```
이전에 리뷰를 놓겠네요.
단순 반복 테스트는 `@RepeatedTest`를 이용하시면 좋습니다.

yamsroun commented on src/test/java/com/megazone/goop/racingcar/car/CarTest.java:
```diff
@@ -13,7 +16,7 @@ class CarTest {
     @BeforeEach
     void setUp() {
-        car = new Car();
+        car = new Car(new CarName("gv80"));
```
저도 갖고 싶습니다!

yamsroun commented on src/test/java/com/megazone/goop/racingcar/car/CarNameTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @DisplayName("자동차 이름은 공백이 아닌 1자 이상 입력해야 한다.")
+    @ParameterizedTest
+    @ValueSource(strings = {"", " "})
```
프로덕션 코드에 체크 로직이 있으므로 `null` 테스트 케이스도 추가하면 좋겠네요.
`@ParameterizedTest`에서 `null` 값은 `@NullSource`로, 빈 문자열은 `@EmptySource`로 대체 가능합니다.
보통은 2가지 함께 필요한 경우가 많으므로 `@NullAndEmptySource`를 사용하셔도 됩니다.

yamsroun commented on src/test/java/com/megazone/goop/racingcar/car/CarNameTest.java:
```diff
@@ -0,0 +1,30 @@
+    void carNameMaxLengthTest(String name) {
+        assertThatThrownBy(() ->
+            new CarName(name)
+        ).isInstanceOf(RuntimeException.class);
```
`isInstanceOf()`는 실제 발생 `Exception`의 상위 타입을 지정해도 통과합니다.
의도적으로 상위 타입을 지정하기도 하지만, 특별한 경우가 아니라면 명확한 `Exception` 타입을 지정하는 것이 좋습니다.

yamsroun commented on src/test/java/com/megazone/goop/racingcar/race/RaceTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @DisplayName("입력한 수만큼 라운드를 반복한다.")
+    @ParameterizedTest
+    @CsvSource(value = {"lay,qm5,k9,sm7:3", "lay,qm5,k9,sm7:5"}, delimiter = ':')
```
Ray가 섭섭해하겠네요.

yamsroun commented on src/test/java/com/megazone/goop/racingcar/race/RaceTest.java:
```diff
@@ -0,0 +1,30 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class RaceTest {
```
`Race` 객체가 제공하는 기능이 몇 가지 더 있으므로, 작은 부분을 테스트하는 여러 TC가 더 필요합니다.

yamsroun reviewed (changes requested):
의식의 흐름대로 리뷰를 진행했습니다.
검토해 주시고, 수정 부탁 드립니다. 🙇🏾‍♀️🙇🏾‍♂️🙇🏾
